File: 백준/Gold/1937. 욕심쟁이 판다/욕심쟁이 판다.java
import java.io.*;
import java.util.*;

/**
 * O(n^3) or O(n^2logn)까지 가능
 * 한 번에 쓸면서 지나가는게 좋을 듯 -> DP
 * d[i][j] : d[i][j]까지 오는데 이동한 최대 거리
 * d[i][j] : Max(d[i - 1][j] d[i][j - 1], d[i + 1][j], d[i][j + 1]) + 1
 * 4방향이라, bottom up으로 채우는 순서가 애매함 -> Top Down
 * 시작은 최댓값에서?
 */


// 그 전 지역보다 옮긴 지역이 대나무가 많아야 함.
// 어디서 출발? 어디로 이동? -> 판다가 이동할 수 있는 칸의 수의 최댓값

public class Main {

    static int n;
    static int[][] board = new int[502][502];
    static int[][] d = new int[502][502];

    static int[] dx = {1, 0, -1, 0};
    static int[] dy = {0, 1, 0, -1};

    public static int dp(int cx, int cy) {
        if (d[cx][cy] != 0) {
            return d[cx][cy];
        }

        for (int i = 0; i < 4; i++) {
            int nx = cx + dx[i];
            int ny = cy + dy[i];

            if (nx < 0 || nx >= n || ny < 0 || ny >= n) continue;
            if (board[cx][cy] <= board[nx][ny]) continue;
            int next = dp(nx, ny);
            d[cx][cy] = Math.max(d[cx][cy], next + 1);
        }

        return d[cx][cy];
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());

        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        int max = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (d[i][j] == 0) {
                    dp(i, j);
                }
                max = Math.max(max, d[i][j]);
            }
        }

        System.out.print(max + 1);
    }

}

File: 백준/Gold/1937. 욕심쟁이 판다/욕심쟁이 판다.kt
/**
 * 대나무를 먹고 자리를 옮기면 그 옮긴 지역에 그 전 지역보다 대나무가 많이 있어야 한다.
 * 목표: "판다가 최대한 많은 칸을 방문할 때 칸수"
 * - 처음에 어디에 배치할 건지
 * - 어디로 이동시킬 건지
 */

import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()
    val arr = Array(n) { StringTokenizer(readLine()).let { st -> IntArray(n) { st.nextToken().toInt() } } }
    val dx = intArrayOf(1, 0, -1, 0)
    val dy = intArrayOf(0, 1, 0, -1)

    val d = Array(n) { IntArray(n) { -1 } } // d[i][j] = (i, j)에서 이동할 수 있는 최대 칸 수

    fun dp(cx: Int, cy: Int): Int {
        if (d[cx][cy] != -1) return d[cx][cy]

        var max = 0
        for (dir in 0 until 4) {
            val nx = cx + dx[dir]
            val ny = cy + dy[dir]

            if (nx !in 0..n-1 || ny !in 0..n-1) continue
            if (arr[nx][ny] <= arr[cx][cy]) continue
            max = maxOf(max, dp(nx, ny))
        }

        d[cx][cy] = max + 1
        return d[cx][cy]
    }

    var max = 0
    for (i in 0 until n) {
        for (j in 0 until n) {
            max = maxOf(max, dp(i, j))
        }
    }

    print(max)
}